components/hab/src/command/plan/render.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2016 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fs::create_dir_all;
+use std::fs::{File, read_to_string};
+use std::io::{Write};
+use std::path::Path;
+use serde_json::{self, Value as Json};
+use toml::Value;
+
+use crate::common::templating::TemplateRenderer;
+use crate::common::ui::{Status, UIWriter, UI};
+use crate::error::Result;
+
+pub fn start(
+    ui: &mut UI,
+    template_path: String,
+    default_toml_path: String,
+    user_toml_path: Option<String>,
+    mock_data_path: Option<String>,
+    print: bool,
+    no_render_dir: bool,
+    render_dir: String,
+    quiet: bool,
+) -> Result<()> {
+    // Strip the file name out of our passed template
+    let file_name = match Path::new(&template_path).file_name() {
+        Some(name) => name.to_str().clone().unwrap(),
+        None => panic!(format!("Something went wrong getting filename of {}", &template_path)),
+    }; 
```

Comment:
Since we're going to use it to construct a path later, there's no need to convert to a `str` and clone: we can leave `file_name` as the `&std::ffi::OsStr` we get back from `Path::file_name`:

```rust
    let file_name = Path::new(&template_path).file_name().expect(&format!(
        "Something went wrong getting filename of {}",
        template_path
    ));
```

this just requires changing how it's used later…

Reply:
I made this

```
Some(name) => Path::new(name)
```

What's the difference in matching like you do with `.expect()` and with `Some/None`?  Is one preferred over the other?

Reply:
```rust
    let file_name = match Path::new(&template_path).file_name() {
        Some(name) => Path::new(name),
        None => panic!(format!(
            "Something went wrong getting filename of {:?}",
            &template_path
        )),
    };
```
is equivalent to
```rust
    let file_name = Path::new(&template_path).file_name().expect(format!(
        "Something went wrong getting filename of {:?}",
        &template_path
    ));
```
I prefer the latter because it's a bit shorter and clearer (to me). Notice there's no `match` in the second version.